src/Interpreters/DDLWorker.cpp

```diff
@@ -775,44 +775,41 @@ bool DDLWorker::tryExecuteQueryOnLeaderReplica(
     String shard_node_name = get_shard_name(task.cluster->getShardsAddresses().at(task.host_shard_num));
     String shard_path = node_path + "/shards/" + shard_node_name;
     String is_executed_path = shard_path + "/executed";
+    String tries_to_execute_path = shard_path + "/tries_to_execute";
     zookeeper->createAncestors(shard_path + "/");
 
-    auto is_already_executed = [&]() -> bool
-    {
-        String executed_by;
-        if (zookeeper->tryGet(is_executed_path, executed_by))
-        {
-            LOG_DEBUG(log, "Task {} has already been executed by leader replica ({}) of the same shard.", task.entry_name, executed_by);
-            return true;
-        }
+    /// Node exists, or we will create or we will get an exception
+    zookeeper->tryCreate(tries_to_execute_path, "0", zkutil::CreateMode::Persistent);
 
-        return false;
-    };
+    static constexpr int MAX_TRIES_TO_EXECUTE = 3;
+
+    String executed_by;
+
+    zkutil::EventPtr event = std::make_shared<Poco::Event>();
+    if (zookeeper->exists(is_executed_path, nullptr, event))
+    {
+        LOG_DEBUG(log, "Task {} has already been executed by replica ({}) of the same shard.", task.entry_name, zookeeper->get(is_executed_path));
+        return true;
+    }
 
     pcg64 rng(randomSeed());
 
     auto lock = createSimpleZooKeeperLock(zookeeper, shard_path, "lock", task.host_id_str);
-    static const size_t max_tries = 20;
+
     bool executed_by_leader = false;
-    for (size_t num_tries = 0; num_tries < max_tries; ++num_tries)
+    while (true)
     {
-        if (is_already_executed())
-        {
-            executed_by_leader = true;
-            break;
-        }
-
         StorageReplicatedMergeTree::Status status;
         replicated_storage->getStatus(status);
 
-        /// Leader replica take lock
+        /// Any replica which is leader tries to take lock
         if (status.is_leader && lock->tryLock())
         {
-            if (is_already_executed())
-            {
-                executed_by_leader = true;
+            /// Doing it exclusively
```

Comment:
Now we don't check that query was already executed after we've got lock. Why?
Looks strange.

Reply:
Very rare case, but nice catch, I'll add it.